### PR TITLE
Added Middleware Parameters

### PR DESCRIPTION
**Pull request by ConnorVG**
This adds middleware parameters (as requested here #6470). The implementation is simple to use.

To pass a parameters to a route middleware, simply do:

``` php
/** etc.. */ [
    'uses' => 'SomeController@method',
    'middleware' => [
        'someMiddleware' => [ /** data to pass */ ],
        'guest' // other middleware works exactly as normal
    ]
] /** etc.. */
```

In a controller:

``` php
$this->middleware('someMiddleware', [ /** options as usual */ ], [ /** data to pass */ ]); 
```

In the Kernal middlewares array:

``` php
protected $middleware = [
    'Illuminate\Foundation\Http\Middleware\CheckForMaintenanceMode',
    'Illuminate\Cookie\Middleware\EncryptCookies',
    'Illuminate\Cookie\Middleware\AddQueuedCookiesToResponse',
    'Illuminate\Session\Middleware\StartSession',
    'Illuminate\View\Middleware\ShareErrorsFromSession',
    'App\Http\Middleware\VerifyCsrfToken',
    'App\Http\Middleware\SomeMiddleware' => [ /** data to pass */ ],
];
```

Accessing the data in the middleware is very simple, it is the third parameter to the handle(...) function:

``` php
public function handle($request, Closure $next, $data = [ /** data that was passed */ ])
{
    return $next($request);
}
```

### CODE REVIEW

File: src/Illuminate/Routing/Route.php [hunk 1]
```diff
@@ -256,7 +256,20 @@ protected function extractOptionalParameters()
 	 */
 	public function middleware()
 	{
-		return (array) array_get($this->action, 'middleware', []);
+		$middleware = (array) array_get($this->action, 'middleware', []);
+
+		foreach ($middleware as $k => $v)
+		{
+			if (!is_numeric($k))
```
Comment (GrahamCampbell) [inline src/Illuminate/Routing/Route.php]: Please fix cs, and also, it would be fine to oneline this if block.

**Comment by GrahamCampbell** [comment]
@taylorotwell should this be targeted at 5.1?

File: src/Illuminate/Routing/Route.php (hunk 1 above)
Comment (ConnorVG) [inline src/Illuminate/Routing/Route.php]: 6833a1a :)

Comment (GrahamCampbell) [inline src/Illuminate/Routing/Route.php]: :+1:

File: src/Illuminate/Pipeline/Pipeline.php [hunk 2]
```diff
@@ -66,7 +66,20 @@ public function send($passable)
 	 */
 	public function through($pipes)
 	{
-		$this->pipes = is_array($pipes) ? $pipes : func_get_args();
+		$pipes = is_array($pipes) ? $pipes : func_get_args();
+
+		foreach ($pipes as $k => $v)
+		{
+			if (!is_numeric($k) || $v instanceof Closure)
```
Comment (GrahamCampbell) [inline src/Illuminate/Pipeline/Pipeline.php]: one liner opportunity again

File: src/Illuminate/Foundation/Http/Kernel.php [hunk 3]
```diff
@@ -126,8 +126,13 @@ public function terminate($request, $response)
 	{
 		$routeMiddlewares = $this->gatherRouteMiddlewares($request);
 
-		foreach (array_merge($routeMiddlewares, $this->middleware) as $middleware)
+		foreach (array_merge($routeMiddlewares, $this->middleware) as $middleware => $v)
 		{
+			if (is_numeric($middleware))
```
Comment (GrahamCampbell) [inline src/Illuminate/Foundation/Http/Kernel.php]: one liner opportunity again

File: src/Illuminate/Pipeline/Pipeline.php (hunk 2 above)
Comment (GrahamCampbell) [inline src/Illuminate/Pipeline/Pipeline.php]: Also, spacing around `!` needs fixing. :)

File: src/Illuminate/Routing/Controller.php [hunk 4]
```diff
@@ -40,11 +40,14 @@
...
 	 * @param  string  $middleware
 	 * @param  array   $options
+	 * @param  array   $data
 	 * @return void
 	 */
-	public function middleware($middleware, array $options = array())
+	public function middleware($middleware, array $options = array(), $data = array())
 	{
-		$this->middleware[$middleware] = $options;
+		$this->middleware[$middleware] = $options + [
```
Comment (GrahamCampbell) [inline src/Illuminate/Routing/Controller.php]: This should be one line.

File: src/Illuminate/Routing/Controller.php [hunk 5]
```diff
@@ -40,11 +40,14 @@
 	 *
 	 * @param  string  $middleware
 	 * @param  array   $options
+	 * @param  array   $data
 	 * @return void
 	 */
-	public function middleware($middleware, array $options = array())
+	public function middleware($middleware, array $options = array(), $data = array())
```
Comment (GrahamCampbell) [inline src/Illuminate/Routing/Controller.php]: Please use short array syntax for new code.

File: src/Illuminate/Routing/Controller.php (hunk 4 above)
Comment (GrahamCampbell) [inline src/Illuminate/Routing/Controller.php]: Also, what about using array_merge?

**Comment by ConnorVG** [comment]
@GrahamCampbell tell me when you've finished commenting and I'll quickly fix them all :)

**Comment by taylorotwell** [comment]
Yes it should be under discussion for 5.1.

**Comment by Burnett01** [comment]
Awesome solution @ConnorVG ! Thank you very much this must be added to 5.1!

**Pull request closed without merging**